### ISSUE TITLE
[LLVM] wrong result from bitcode compiled from C source

### ISSUE THREAD

**Issue by zhendongsu** [oracle/graal#4132]
GraalVM version or commit id if built from source: GraalVM CE 22.1.0-dev-20211215_1015 (the latest snapshot build from https://github.com/graalvm/graalvm-ce-dev-builds/releases)
CE or EE: CE
JDK version: openjdk 11.0.13
OS and OS Version: Ubuntu 18.04
Architecture: amd64
The output of `java -Xinternalversion`:
OpenJDK 64-Bit Server VM (11.0.13+8-jvmci-22.0-b02) for linux-amd64 JRE (11.0.13+8-jvmci-22.0-b02), built on Nov 22 2021 22:03:35 by "buildslave" with gcc 7.3.0

```
[532] % clangg22dev -v
GraalVM wrapper script for clang
GraalVM version: Development Build
running: /local/suz-local/software/graalvm-ce-java11-22.1.0-dev/lib/llvm/bin/clang -v
clang version 12.0.1 (GraalVM.org llvmorg-12.0.1-3-g6e0a5672bc-bgf11ed69a5a 6e0a5672bc058d882dce3d56f90b72b64a6870d7)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/software/graalvm-ce-java11-22.1.0-dev/lib/llvm/bin
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda, version 11.0
[533] % lli22dev --version
LLVM 12.0.1 (GraalVM CE Native 22.1.0-dev)
[534] % opt22dev --version
LLVM (http://llvm.org/):
  LLVM version 12.0.1-3-g6e0a5672bc-bgf11ed69a5a
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: broadwell
[535] % 
[535] % clangg22dev -O1 -c -emit-llvm -o small.bc small.c
[536] % opt22dev --loop-extract --attributor -o small-opt.bc small.bc
[537] % 
[537] % clangg22dev -fno-lto small-opt.bc -o small.exe
.[538] % ./small.exe
[539] % 
[539] % timeout -s 9 1 lli22dev small-opt.bc >& wrong-output.txt
Killed
[540] % wc wrong-output.txt
      0       1 1949696 wrong-output.txt
[541] % 
[541] % cat small.c
int printf(const char *, ...);
int a = 1, b;
int main() {
  int c = a ^ 2, e;
  if (c) {
    c = a;
    if (b)
      goto L;
  }
  int f = b | c, g = c;
 L:
  while (c > 1) {
    int h[2] = {e, e};
  }
  if (!f && g) {
    printf("%d", 0);
    goto L;
  }
  if (!g)
    goto L;
  return 0;
}
```

Expected output: `lli22dev small-opt.bc` should return normally and produce no output, rather than running into an infinite loop.
Note: It might be related to https://github.com/oracle/graal/issues/4131.

**Comment by oubidar-Abderrahim** [comment]
Hi, Thank you for reporting this, we will take a look into it and get back to you

**Comment by rudihorn** [comment]
Hi @zhendongsu, thanks for reporting this. This is actually an upstream bug in LLVM 12.0.1. I had a look at the produced `small-opt.bc` file, and the compiler introduces undefined behaviour for the following phi in `main.extracted.1`:

```
%5 = phi i32 [ undef, %newFuncRoot ], [ %8, %._crit_edge7 ]
```

The bug can be reproduced without Graal by compiling `small-opt.bc` with clang 13. It can also be fixed by performing the optimization step with `opt` version 13 instead. The same phi does not have undefined behaviour anymore:

```
%5 = phi i32 [ %1, %newFuncRoot ], [ %8, %11 ]
```

Could you maybe confirm this is the same issue you were facing?

**Comment by rschatz** [comment]
This should be fixed with our recent update to LLVM 14.0.3 (https://github.com/oracle/graal/commit/7c64fd84dd2a31f83460ed4c948a9d7d9c628a09).
